src/drivers/batt_smbus/batt_smbus.cpp

```diff
@@ -274,28 +84,58 @@ BATT_SMBUS::~BATT_SMBUS()
 	if (_manufacturer_name != nullptr) {
 		delete[] _manufacturer_name;
 	}
+
+	// Switch back to ADC battery measurement
+	PX4_ERR("Failed to start smart battery. Switching to ADC measurement.");
+	int battsource = 0;
+	param_set(param_find("BAT_SOURCE"), &battsource);
 }
 
 int
 BATT_SMBUS::init()
 {
-	int ret = ENOTTY;
+	int ret = PX4_ERROR;
 
-	// attempt to initialise I2C bus
+	// attempt to initialize I2C bus
 	ret = I2C::init();
 
 	if (ret != OK) {
-		PX4_ERR("failed to init I2C");
+		PX4_ERR("Failed to init I2C");
 		return ret;
+	}
 
-	} else {
-		//Find the battery on the bus
-		search();
+	// Find the battery on the bus and read startup info
+	ret = search();
+
+	if (ret != OK) {
+		return ret;
+	}
 
-		// start work queue
-		start();
+	// Retry up to 10 times to read startup info
+	for (size_t i = 0; i < 10; i++) {
+		ret = GetStartupInfo();
+
+		if (ret == OK) {
+			break;
+		}
+	}
+
+	if (ret != OK) {
+		PX4_ERR("Failed to get battery startup info");
+		return ret;
 	}
 
+	int battsource = 0;
+	param_get(param_find("BAT_SOURCE"), &battsource);
+
+	if (battsource != 1) {
+		battsource = 1;
+		param_set(param_find("BAT_SOURCE"), &battsource);
```

Comment:
Why are you setting the parameter?

Reply:
If the driver starts successfully, it sets batt_source to bypass the ADC battery updates. If the driver fails to start or gets stopped, it will set batt_source to 0 to use the ADC measurement.

Reply:
BAT_SOURCE is the users desired configuration, not something that should be set to indicate the presence of something.